docs/couler-api-design.md

```diff
@@ -0,0 +1,87 @@
+# Couler API Design
+
+This document outlines the design of core Couler APIs to support multiple workflow backends.
+
+## Goals
+
+* Design the core Couler APIs that can be implemented across different workflow engines.
+* Provide a *minimal* set of APIs that are forward-looking, engine-agnostic, and less likely to deprecate over-time.
+
+## Non-Goals
+
+* Provide an exhaustive set of APIs that cover all use cases and all workflow engines.
+* Provide implementation details.
+
+## Design
+
+Core operations (`couler.ops`):
+
+* `run_step(step_def)` where `step_def` is the step definition that can be a container spec, Python function,
+or spec that's specific to the underlying workflow engine (e.g. k8s CRD if Argo Workflow is used).
+
+Control flow (`couler.control_flows`):
+
+* `map(func, *args, **kwargs)` where `*args` and `**kwargs` correspond to various arguments passed to the function `func`.
+* `when(cond, if_op, else_op)` where
+    * `cond` can be any of the following predicates: `equal()`, `not_equal()`, `bigger()`, `smaller()`, `bigger_equal()`, `smaller_equal()`.
+    * The operation defined in `if_op` will be executed when `cond` is true. Otherwise, `else_op` will be executed.
```

Comment:
are we missing a `next` op that simply specifies the order of steps? (e.g. a linear workflow)

Reply:
Actually it was intentional since Couler's APIs are imperative so users can define the op and once it's declared it will be added automatically to the workflow as one of the steps. Users can also set the dependencies explicitly to construct a linear workflow if they want.

Reply:
Should we add `dag` and `set_dependencies` in this doc?

Based on the current `dag` design, if users want to add a `when` judge, they can only add `couler.when` outside of the `couler.dag`. What's the best practice of this case?